### ISSUE TITLE
Twitter and Facebook card look and info

### ISSUE THREAD

**Issue by smit1678** [hotosm/tasking-manager#1035]
Currently we only expose the description and overall site title to Twitter cards and Facebook previews: 

![10 twitter 2018-01-26 14-43-38](https://user-images.githubusercontent.com/796838/35444724-54589368-02a7-11e8-9d06-c7d82f948b5c.png)

It might be nice and useful when a task is shared that there is engaging information or look on the preview. The goal would be to provide more information about the current task or a visual that takes advantage of the preview space Twitter and Facebook offers. 

What would be useful information to add? Are we constrained by the frontend architecture on what we can show? 

****

From the [Twitter card validator](https://cards-dev.twitter.com/validator):

![card validator twitter developers 2018-01-26 14-39-38](https://user-images.githubusercontent.com/796838/35444626-0e207c08-02a7-11e8-9414-d121830e9c40.png)

From the [Facebook Debugger](https://developers.facebook.com/tools/debug/sharing/?q=https%3A%2F%2Ftasks.hotosm.org%2Fproject%2F3744): 

![sharing debugger - facebook for developers 2018-01-26 14-42-10](https://user-images.githubusercontent.com/796838/35444647-18bdfda2-02a7-11e8-8955-fbca062e322e.png)

**Comment by bgirardot** [comment]
@smit1678 Very helpful information. Could you add it to the existing issue #776 ?

**Comment by bgirardot** [comment]
@smit1678 There is also one for twitter issue #777

**Comment by smit1678** [comment]
I thought I did a proper search for previous issues. 

We should either close this out and rewrite the others or close the others out and use this one.

**Comment by bgirardot** [comment]
I linked this one to the existing issues so am going to close it.

**Comment by TheChirpyWitch** [comment]
Hey @smit1678 and @bgirardot ! I am trying to contribute to this project and just to understand the nature of the problem here, you want the facebook/twitter sharing card to a particular task's description, associated image, etc. right?

**Comment by bgirardot** [comment]
@TheChirpyWitch For sure we would like the title of the project and some text from the Description field (Short description often has some tasking manager specific text in it) instead of the generic HOT Tasking Manager description. Also cool might be the %Completed. Not sure about a graphic, we do not currently have a spot for a graphic specifically. but it would be great to have something, even if it was just one image for all the tweets/fb.

**Comment by TheChirpyWitch** [comment]
Got it. I'll work on this issue then. :)

**Comment by k3KAW8Pnf7mkmdSMPHz27** [comment]
I am aware that this doesn't seem to be a high priority issue, but if no one else is working on it, I'd love to give it a shot.

I have had some problems getting everything up and going (I intend to open up an issue on it next week with more details), but I'll add a draft/WIP PR as soon as I have something more concrete.

**Comment by k3KAW8Pnf7mkmdSMPHz27** [comment]
I am not really convince this is should be tagged as `Difficulty 1`.

The TL;DR is that right now neither Facebook nor Twitter supports updating the `<meta ...>` tags with JavaScript for the preview in cards.
The most common solutions are either [pre-rendering content](https://create-react-app.dev/docs/pre-rendering-into-static-html-files) or [server-side updating of the tags](https://create-react-app.dev/docs/title-and-meta-tags/#generating-dynamic-meta-tags-on-the-server). [There were a similar discussion regarding the hrmi-dataportal](https://github.com/HumanRightsMeasurementInitiative/hrmi-dataportal/issues/162).

The current `<meta>` tags can be improved with pre-rendering but without knowing what serves the SPA it would be hard to keep the projects up to date with this approach.

I am currently assuming server-side rendering is not an option.

**Comment by willemarcel** [comment]
@k3KAW8Pnf7mkmdSMPHz27 feel free to propose a solution. You're right, SSR is not an option.
We use that library to update the title tags of each page https://github.com/rstacruz/react-meta-elements, maybe it can solve the problem.

**Comment by k3KAW8Pnf7mkmdSMPHz27** [comment]
`react-meta-elements` can definitively be part of a solution. The only issue is that it does not update/delete meta tags by default, but that should be possible to work around.

As for solutions, some sort of pre-rendering must be done. At the moment, I am looking into [react-snap](https://github.com/stereobooster/react-snap) as it seems to be the more maintained [pre-render/snapshot solution mentioned in the create-react-app docs](https://create-react-app.dev/docs/pre-rendering-into-static-html-files), and it does not require any Webpack configuration changes. It produces the correct title and meta-tags for `index.html`, and unfortunately, some strange visual glitches. I'll need a bit more time to fiddle around with the configurations to see if the issues disappear. If they don't, I'll try some other pre-render technique 😃 
My current goal is to try to get `/about` (a page with very few dynamic elements) working with Slack preview (which has the same meta-tag issue as Twitter/Facebook but benefits from a solution using [oembed](https://oembed.com/), which is likely useful to #3554). 

If there is something that isn't clear, feel free to ask. I am new to this subject so, if I am doing something incredibly stupid, there might be something obvious that I have missed 😱 

"P.S."
I am also assuming that using a paid third-party service is not an option 😛

**Comment by willemarcel** [comment]
@k3KAW8Pnf7mkmdSMPHz27 It's a nice proposal. As an open source software, we try to keep it simple to other organizations to deploy, so we avoid third-party services. I've pushed some little improvements on #3590

**Comment by k3KAW8Pnf7mkmdSMPHz27** [comment]
@willemarcel thank you for your comments (and the link), I'll be using it while debugging.

> we try to keep it simple to other organizations to deploy

This clarifies the restrictions to me. For now, I am mostly attempting to get something working, but I'll see if I can find a solution that is easier to debug and less complex than react-snap.

**Comment by k3KAW8Pnf7mkmdSMPHz27** [comment]
The nutshell version is that the `react-snap` method can work, but it comes with some limitations,
1. The server must look for `index.html` files based on the URL (e.g., `https://tasks.hotosm.org/projects/8863` -> `https://tasks.hotosm.org/projects/8863/index.html`). This appears to be the default on most static web hosts (including S3). It works with `serve` (but not with the `-s` option). You can see this redirect in action in the card validator images below.
2. All dynamic pages must be reachable through crawling. You can set several starting roots (e.g., `/` and `/api-docs/`).
3. As all information information will be in static `index.html` files, they have to be updated with new information. Which means, crawl the webpage with `react-snap` and then sync/upload the new/changed `index.html` files with the static web host.

As 3. in particular will make it harder to deploy TM (and is as far as I can tell unavoidable), this might be a good time for me to stop looking at this issue.
If, despite these limitations, this is something TM wants, I can look into how to
1. make this an opt-in feature in `tasking-manager.env` (similar to service workers)
2. integrate it into the AWS setup (e.g., crawl and update the site daily/weekly etc.)


<img width="995" alt="Skärmavbild 2020-09-21 kl  17 02 59" src="https://user-images.githubusercontent.com/2598631/93821642-d5908400-fc2c-11ea-8eb6-ea6318adcd37.png">
<img width="995" alt="Skärmavbild 2020-09-21 kl  17 02 48" src="https://user-images.githubusercontent.com/2598631/93821653-d7f2de00-fc2c-11ea-86c1-f2ddf8922475.png">
"P.S."
I used the organization image as it is intended as a proof-of-concept. Perhaps the zoomed in map-view could be used.

**Comment by willemarcel** [comment]
@k3KAW8Pnf7mkmdSMPHz27 thanks for the awesome research! Yeah, I think the effort to make it possible would be too big for a non critical issue...

**Comment by k3KAW8Pnf7mkmdSMPHz27** [comment]
@willemarcel that seems likely. I appreciate that you have taken the time to respond on a low priority issue!
